design/ux/Obscuro_Gateway.md

```diff
@@ -0,0 +1,264 @@
+- Return UserId
+
+Note: Has to be protected against DDOS attacks.
+
+### Authenticate address - POST "/authenticate?u=$UserId"
+JSON Fields:
+- text
+- signature
+
+This call will be made by a javascript function after it has collected the signed text containing the UserId and the Address from the wallet.
+
+This call is equivalent to the current: "submitviewingkey/", but instead it will save the information against the UserId.
+
+Actions:
+- check the text is well formed and extract the userId and address
+- check the UserId corresponds to the one in the text
+- check the signature corresponds to the address and is valid
+- save the text+signature against the userId 
+
+### ETH RPC endpoints
+All the Eth RPC endpoints are implemented as they are now in the WE. 
+The difference is that the UserId must be checked before any logic, and the registered addresses for that user are loaded in context.
+
+## SGX
+
+- Investigate eGo and how https connections can be terminated in an enclave.
+
+
+## Javascript - UI
+
+The "3-click" flow must be implemented as in the above diagram.
+The first UI can be ugly.
+
+Note that the current WE implements most of this flow.
+
+
+## Upgradability
+
+When the OG is upgraded, the database with viewing keys has to be handed over to the new version.
+The best mechanism is to use the transparent upgrading approval we have for the enclave based on an event from the Management Contract.
+
+
+# Tasks
+The tasks can be split up in 3 streams: Implementing multi-tenancy, SGX and the UI. 
+
+### Multi tenancy stream
+
+I propose to start gradually by adding functionality that doesn't break the WE.
+
+1. Implement an SQL database. 
+   The tests can use sqlite, same as we do for the enclave. And in the real setup it will be edgelessdb. 
+   This code can be extracted from the enclave and made reusable.
+
+2. Store the VKs in the database against a hardcoded userId and create the logic to fetch them based on the hardcoded userId.
+
+Note: At this stage there is no functional change of the wallet extension. Just prepararation. 
+
+3. Add the UserId parameter in all request handlers.
+
+Note: still no functional change, except that now all the rpc urls in the test need "?u=1" to work
+
+4. Change the format of the signed string and implement the authenticate endpoint.
+
+5. Do the rest of the wiring of the userId
+
+6. Implement the "join" endpoint, and make sure everything works E2E
+
+
+### SGX stream
+
+1. The dev-ops step: Building the code using the ego tooling. The "enclave" can be used as a model for this.
+2. HTTPS connection ending in the enclave
+
+
+### UI stream
+Very simple interface to show that the 3-click approach is possible.
+
+
+# Advanced features
+
+## Allowing users to create accounts on the OG
+Maybe using single-sign On.
+
+This will allow advanced features.
+
+## Empowering other addresses to view some data. 
+
+This could be used for tax purposes, or to prove holdings.
+
+## Revocation of userId 
```

Comment:
Would be good to describe here rotation along with what happens with rewind attacks and upgrades; I doubt there is anything major there but it sounds like a potential vector

Reply:
what sort of rewind attack do you have in mind?

Reply:
Just the general one where some userID leaks and the operator rewinds the state back to when it had a viewing key and registering another account in order to browse up to date info of the user that can be found in the actual node.

Reply:
yeah. that would work.
I'll add a note